File: sort/quick-sort.js
let unsorted = [5, 2, 9, 1, 11, 7, 8, -1, 4, 3, 10, 14, 0, 6];

const quick_sort = (unsorted, start, end) => {

    if (start < end) {

        let pivot = splitArray(unsorted, start, end);

        quick_sort(unsorted, start, pivot - 1);

        quick_sort(unsorted, pivot + 1, end);
    }

    return unsorted;
}

const splitArray = (unsorted, start, end) => {

    for (let i = start; i <= end; i++) {

        if (unsorted[i] <= unsorted[end]) {

            if (unsorted[i] != unsorted[start]) {

                let temp = unsorted[i];
                unsorted[i] = unsorted[start];
                unsorted[start] = temp;

            }

            start++;
        }
    }

    return start - 1;
}

console.log(quick_sort(unsorted, 0, unsorted.length - 1));

File: sort/QuickSort.java
package Sort;
public class QuickSort {

    public static void main(String[] args) {
        
        int[] test = { 3, 5, 1, 6, 2, 7, 4, 8, 9 };

        quickSort(test, 0, test.length - 1);

        for (int i : test) {
            System.out.println(i);
        }
    }

    public static void quickSort(int[] unsorted, int low, int high) {

        if (low < high) {
            int pivot = partition(unsorted, low, high);
            quickSort(unsorted, low, pivot - 1);
            quickSort(unsorted, pivot + 1, high);
        }
    }
    
    public static int partition(int[] unsorted, int left, int right) {

        int i = left;
        int j = right;

        while (i < j) {

            while (i < j && unsorted[j] >= unsorted[left]) {
                j--;
            }

            while (i < j && unsorted[i] <= unsorted[left]) {
                i++;
            }

            int temp = unsorted[i];
            unsorted[i] = unsorted[j];
            unsorted[j] = temp;
        }
        
        int temp = unsorted[left];
        unsorted[left] = unsorted[i];
        unsorted[i] = temp;

        return i;
    }
}